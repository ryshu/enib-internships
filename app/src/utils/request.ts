import axios from 'axios';
import { Message, MessageBox } from 'element-ui';

const service = axios.create({
  baseURL: process.env.VUE_APP_BASE_API, // url = base url + request url
  timeout: 5000,
  // withCredentials: true // send cookies when cross-domain requests
});

// Request interceptors
service.interceptors.request.use(
  config => config,
  error => {
    Promise.reject(error);
  }
);

// Response interceptors
service.interceptors.response.use(
  response => {
    const res = response.data;
    if (res.code && res.code !== 200 && res.code !== 201) {
      Message({
        message: res.message || 'Error',
        type: 'error',
        duration: 5 * 1000,
      });
      if (res.code === 500) {
        MessageBox.confirm(
          '你已被登出，可以取消继续留在该页面，或者重新登录',
          '确定登出',
          {
            confirmButtonText: '重新登录',
            cancelButtonText: '取消',
            type: 'warning',
          }
        ).then(() => {
          window.location.assign('/logout');
        });
      }
      return Promise.reject(new Error(res.message || 'Error'));
    } else {
      return response.data;
    }
  },
  error => {
    console.log(error);
    Message({
      message: error.message,
      type: 'error',
      duration: 5 * 1000,
    });
    return Promise.reject(error);
  }
);

export default service;
