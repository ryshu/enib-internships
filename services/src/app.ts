import express, { Request, Response, NextFunction } from 'express';
import bodyParser from 'body-parser';
import cors from 'cors';
import compression from 'compression'; // compresses requests
import lusca from 'lusca';
import dotenv from 'dotenv';
import flash from 'express-flash';
import path from 'path';
import methodOverride from 'method-override';

// Load environment variables from .env file, where API keys and passwords are configured
dotenv.config({ path: '.env' });

// Loggers
import logger from './utils/logger';
import { APIError } from './utils/error';
import httpStatus from 'http-status-codes';

// router and database
import router from './configs/setup/route';
import './configs/instances/database'; // Only import to setup
import './configs/setup/database'; // Only import to setup

// Create Express server
const app = express();

// Express configuration
app.set('port', process.env.INTERNSHIP_ENIB_API_PORT || 3000);
app.use(compression());
app.use(bodyParser.json({ limit: '1mb' }));
app.use(bodyParser.urlencoded({ extended: true, limit: '1mb' }));
app.use(flash());
app.use(lusca.xframe('SAMEORIGIN'));
app.use(lusca.xssProtection(true));

app.use(express.static(path.join(__dirname, 'public'), { maxAge: 31557600000 }));

const corsOptions = {
    origin: 'http://localhost:4200',
    optionsSuccessStatus: 200,
};
app.use(cors(corsOptions));

app.use(router);

/**
 * Error handler setup
 */
app.use(methodOverride());
app.use((err: APIError | Error, _req: Request, res: Response, _next: NextFunction) => {
    if (err instanceof APIError) {
        return res.status(err.status).json(err);
    } else {
        logger.error(err);
        return res.status(httpStatus.INTERNAL_SERVER_ERROR).send(`Internal server error`);
    }
});

export default app;
