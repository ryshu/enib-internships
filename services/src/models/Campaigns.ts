import * as Sequelize from 'sequelize';

import database from '../configs/instances/database';

import MPS from './MentoringPropositions';
import Internships from './Internships';

class Campaigns extends Sequelize.Model implements ICampaignEntity {
    public static associations: {
        propositions: Sequelize.Association<Campaigns, MPS>;
        availbleInternships: Sequelize.Association<Campaigns, Internships>;
        validatedInternships: Sequelize.Association<Campaigns, Internships>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public name: string;
    public description: string;
    public startAt: number;
    public endAt: number;
    public semester: string;
    public maxProposition: number;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public getPropositions: Sequelize.HasManyGetAssociationsMixin<MPS>;
    public addProposition: Sequelize.HasManyAddAssociationMixin<MPS, MPS['id']>;
    public createProposition: Sequelize.HasManyCreateAssociationMixin<IMentoringPropositionEntity>;
    public hasProposition: Sequelize.HasManyHasAssociationMixin<MPS, MPS['id']>;
    public countPropositions: Sequelize.HasManyCountAssociationsMixin;

    // AvailableInternships
    public getAvailableInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addAvailableInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createAvailableInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasAvailableInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countAvailableInternships: Sequelize.HasManyCountAssociationsMixin;

    // ValidatedInternships
    public getValidatedInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addValidatedInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createValidatedInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasValidatedInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countValidatedInternships: Sequelize.HasManyCountAssociationsMixin;

    public readonly propositions?: MPS[];
    public readonly availableInternships?: Internships[];
    public readonly validatedInternships?: Internships[];
}

Campaigns.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        name: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        description: {
            type: new Sequelize.DataTypes.STRING(),
            allowNull: false,
        },
        startAt: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            allowNull: true,
            defaultValue: null,
        },
        endAt: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            allowNull: true,
            defaultValue: null,
        },
        semester: {
            type: Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        maxProposition: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            defaultValue: 0,
            allowNull: true,
        },
    },
    {
        tableName: 'campaigns',
        sequelize: database,
    },
);

export default Campaigns;
