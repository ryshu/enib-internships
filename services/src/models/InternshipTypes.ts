import * as Sequelize from 'sequelize';

import database from '../configs/instances/database';

import Internships from './Internships';

class InternshipTypes extends Sequelize.Model implements IInternshipTypeEntity {
    public static associations: {
        internships: Sequelize.Association<InternshipTypes, Internships>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public label: string;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public getInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countInternships: Sequelize.HasManyCountAssociationsMixin;

    public readonly internships?: Internships[];
}

InternshipTypes.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        label: {
            type: new Sequelize.DataTypes.TEXT(),
            allowNull: true,
        },
    },
    {
        tableName: 'internship-types',
        sequelize: database,
    },
);

export default InternshipTypes;
