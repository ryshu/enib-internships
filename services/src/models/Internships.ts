import * as Sequelize from 'sequelize';

import database from '../configs/instances/database';

import Businesses from './Businesses';
import InternshipTypes from './InternshipTypes';

class Internships extends Sequelize.Model implements IInternshipEntity {
    public static associations: {
        category: Sequelize.Association<Internships, InternshipTypes>;
        business: Sequelize.Association<Internships, Businesses>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    // Data
    public subject: string;
    public description: string;

    // Localisation
    public country: string;
    public city: string;
    public postalCode: string;
    public address: string;
    public additional?: string;

    // State
    public isInternshipAbroad: boolean;
    public isValidated: boolean;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public getBusiness: Sequelize.BelongsToGetAssociationMixin<Businesses>;
    public setBusiness: Sequelize.BelongsToSetAssociationMixin<Businesses, Businesses['id']>;
    public createBusiness: Sequelize.BelongsToCreateAssociationMixin<IBusinessEntity>;

    public getCategory: Sequelize.BelongsToGetAssociationMixin<InternshipTypes>;
    public setCategory: Sequelize.BelongsToSetAssociationMixin<
        InternshipTypes,
        InternshipTypes['id']
    >;
    public createCategory: Sequelize.BelongsToCreateAssociationMixin<IInternshipTypeEntity>;

    public readonly business?: Businesses | Businesses['id'];
    public readonly category?: InternshipTypes | InternshipTypes['id'];
}

Internships.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },

        // Data
        subject: {
            type: new Sequelize.DataTypes.STRING(256),
            allowNull: false,
        },
        description: {
            type: new Sequelize.DataTypes.TEXT(),
            allowNull: true,
        },

        // Localisation
        country: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        city: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        postalCode: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        address: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        additional: {
            type: new Sequelize.DataTypes.STRING(),
            allowNull: true,
        },

        // State
        isInternshipAbroad: {
            type: Sequelize.DataTypes.BOOLEAN,
            allowNull: false,
            defaultValue: false,
        },
        isValidated: {
            type: Sequelize.DataTypes.BOOLEAN,
            allowNull: false,
            defaultValue: false,
        },
    },
    {
        tableName: 'internships',
        sequelize: database,
        defaultScope: {
            attributes: { exclude: ['businessId', 'categoryId'] },
        },
    },
);

export default Internships;
