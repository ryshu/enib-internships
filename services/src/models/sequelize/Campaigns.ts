import * as Sequelize from 'sequelize';

import database from '../../configs/instances/database';

import MPS from './MentoringPropositions';
import Mentors from './Mentors';
import InternshipTypes from './InternshipTypes';
import Internships from './Internships';

import {
    IInternshipEntity,
    IMentoringPropositionEntity,
    IInternshipTypeEntity,
} from '../../declarations';

class Campaigns extends Sequelize.Model {
    public static associations: {
        propositions: Sequelize.Association<Campaigns, MPS>;
        availbleInternships: Sequelize.Association<Campaigns, Internships>;
        validatedInternships: Sequelize.Association<Campaigns, Internships>;
        mentors: Sequelize.Association<Campaigns, Mentors>;
        category: Sequelize.Association<Campaigns, InternshipTypes>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public name: string;
    public description: string;

    public semester: string;
    public maxProposition: number;

    public isPublish: boolean;

    public startAt: number;
    public endAt: number;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    // Proposition
    public getPropositions: Sequelize.HasManyGetAssociationsMixin<MPS>;
    public addProposition: Sequelize.HasManyAddAssociationMixin<MPS, MPS['id']>;
    public createProposition: Sequelize.HasManyCreateAssociationMixin<IMentoringPropositionEntity>;
    public hasProposition: Sequelize.HasManyHasAssociationMixin<MPS, MPS['id']>;
    public countPropositions: Sequelize.HasManyCountAssociationsMixin;

    // Mentor
    public getMentors: Sequelize.BelongsToManyGetAssociationsMixin<Mentors>;
    public addMentor: Sequelize.BelongsToManyAddAssociationMixin<Mentors, Mentors['id']>;
    public hasMentor: Sequelize.BelongsToManyHasAssociationMixin<Mentors, Mentors['id']>;
    public countMentors: Sequelize.BelongsToManyCountAssociationsMixin;

    // Category
    public getCategory: Sequelize.BelongsToGetAssociationMixin<InternshipTypes>;
    public setCategory: Sequelize.BelongsToSetAssociationMixin<
        InternshipTypes,
        InternshipTypes['id']
    >;
    public createCategory: Sequelize.BelongsToCreateAssociationMixin<IInternshipTypeEntity>;

    // AvailableInternships
    public getAvailableInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addAvailableInternship: Sequelize.HasManyAddAssociationMixin<
        Internships,
        Internships['id']
    >;
    public createAvailableInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasAvailableInternship: Sequelize.HasManyHasAssociationMixin<
        Internships,
        Internships['id']
    >;
    public countAvailableInternships: Sequelize.HasManyCountAssociationsMixin;

    // ValidatedInternships
    public getValidatedInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addValidatedInternship: Sequelize.HasManyAddAssociationMixin<
        Internships,
        Internships['id']
    >;
    public createValidatedInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasValidatedInternship: Sequelize.HasManyHasAssociationMixin<
        Internships,
        Internships['id']
    >;
    public countValidatedInternships: Sequelize.HasManyCountAssociationsMixin;

    public readonly propositions?: MPS[];
    public readonly availableInternships?: Internships[];
    public readonly validatedInternships?: Internships[];
    public readonly mentors?: Mentors[];
    public readonly category: InternshipTypes;
}

Campaigns.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        name: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        description: {
            type: new Sequelize.DataTypes.TEXT(),
            allowNull: false,
        },

        semester: {
            type: Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        maxProposition: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            defaultValue: 0,
            allowNull: true,
        },

        isPublish: {
            type: Sequelize.DataTypes.BOOLEAN,
            defaultValue: false,
        },

        startAt: {
            type: Sequelize.DataTypes.BIGINT,
            allowNull: true,
            defaultValue: null,
        },
        endAt: {
            type: Sequelize.DataTypes.BIGINT,
            allowNull: true,
            defaultValue: null,
        },
    },
    {
        tableName: 'campaigns',
        sequelize: database,
        defaultScope: {
            attributes: { exclude: ['categoryId'] },
        },
    },
);

export default Campaigns;
