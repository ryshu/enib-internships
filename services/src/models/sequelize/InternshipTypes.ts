import * as Sequelize from 'sequelize';

import database from '../../configs/instances/database';

import Internships from './Internships';
import Campaigns from './Campaigns';

import { ICampaignEntity, IInternshipEntity } from '../../declarations';

class InternshipTypes extends Sequelize.Model {
    public static associations: {
        internships: Sequelize.Association<InternshipTypes, Internships>;
        campaigns: Sequelize.Association<InternshipTypes, Campaigns>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public label: string;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    // Internships
    public getInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countInternships: Sequelize.HasManyCountAssociationsMixin;

    // Campaigns
    public getCampaigns: Sequelize.HasManyGetAssociationsMixin<Campaigns>;
    public addCampaign: Sequelize.HasManyAddAssociationMixin<Campaigns, Campaigns['id']>;
    public createCampaign: Sequelize.HasManyCreateAssociationMixin<ICampaignEntity>;
    public hasCampaign: Sequelize.HasManyHasAssociationMixin<Campaigns, Campaigns['id']>;
    public countCampaigns: Sequelize.HasManyCountAssociationsMixin;

    public readonly internships?: Internships[];
    public readonly campaigns?: Campaigns[];
}

InternshipTypes.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        label: {
            type: new Sequelize.DataTypes.TEXT(),
            allowNull: true,
        },
    },
    {
        tableName: 'internship-types',
        sequelize: database,
        timestamps: true,
        paranoid: true,
    },
);

export default InternshipTypes;
