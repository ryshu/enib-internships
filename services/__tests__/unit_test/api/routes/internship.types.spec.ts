import request from 'supertest';

// Include app for super agent, not server
import app from '../../../../src/app';

// Include database connection
import dbSetup from '../../../../src/configs/setup/database';

// Import model for pre-operation before asserting API methods
import InternshipTypes from '../../../../src/models/sequelize/InternshipTypes';

import {
    defaultInternshipTypes,
    defaultInternships,
    defaultCampaigns,
} from '../../../../__mocks__/mockData';

import Internships from '../../../../src/models/sequelize/Internships';
import Campaigns from '../../../../src/models/sequelize/Campaigns';

const baseURL = `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}`;

jest.setTimeout(30000);

beforeAll((done) => {
    dbSetup.then(() => done()).catch((e) => done(e));
});

describe('GET /internshipTypes', () => {
    beforeEach(async () => {
        // Remove all
        await InternshipTypes.destroy({ where: {} });
    });

    it('NoInternshipTypes_204', async () => {
        const RESPONSE = await request(app).get(`${baseURL}/internshipTypes`);
        expect(RESPONSE.status).toBe(204);
    });

    it('InternshipTypes_200', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();

        await InternshipTypes.create(VALID_INTERNSHIP_TYPES);
        const RESPONSE = await request(app).get(`${baseURL}/internshipTypes`);
        expect(RESPONSE.status).toBe(200);
        expect(Array.isArray(RESPONSE.body)).toBeTruthy();

        expect(RESPONSE.body).toBeTruthy();
        expect(RESPONSE.body).toHaveLength(1);
    });
});

describe('POST /internshipTypes', () => {
    it('NoBody_400', async () => {
        const RESPONSE = await request(app).post(`${baseURL}/internshipTypes`);
        expect(RESPONSE.status).toBe(400);
    });

    it('ValidInternshipTypes_200', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();

        const RESPONSE = await request(app)
            .post(`${baseURL}/internshipTypes`)
            .send(VALID_INTERNSHIP_TYPES);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('GET /internshipTypes/:id', () => {
    beforeEach(async () => {
        // Remove all
        await InternshipTypes.destroy({ where: {} });
    });

    it('NoInternshipTypes_204', async () => {
        const RESPONSE = await request(app).get(`${baseURL}/internshipTypes/10`);
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).get(`${baseURL}/internshipTypes/{falseEncoding}`);
        expect(RESPONSE.status).toBe(400);
    });

    it('InternshipTypes_200', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();

        const CREATED = await InternshipTypes.create(VALID_INTERNSHIP_TYPES);
        const RESPONSE = await request(app).get(`${baseURL}/internshipTypes/${CREATED.id}`);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('PUT /internshipTypes/:id', () => {
    beforeEach(async () => {
        // Remove all
        await InternshipTypes.destroy({ where: {} });
    });

    it('NoInternshipTypes_204', async () => {
        const RESPONSE = await request(app).put(`${baseURL}/internshipTypes/10`);
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).put(`${baseURL}/internshipTypes/{falseEncoding}`);
        expect(RESPONSE.status).toBe(400);
    });

    it('InternshipTypes_200_UpdateAllData', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();

        const CREATED = await InternshipTypes.create(VALID_INTERNSHIP_TYPES);

        // Change some data on VALID_INTERNSHIP_TYPES
        VALID_INTERNSHIP_TYPES.label = 'additional data';

        const RESPONSE = await request(app)
            .put(`${baseURL}/internshipTypes/${CREATED.id}`)
            .send(VALID_INTERNSHIP_TYPES);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });

    it('InternshipTypes_200_NotAnyDataUpdated', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();

        const CREATED = await InternshipTypes.create(VALID_INTERNSHIP_TYPES);

        const RESPONSE = await request(app)
            .put(`${baseURL}/internshipTypes/${CREATED.id}`)
            .send({});
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('DELETE /internshipTypes/:id', () => {
    beforeEach(async () => {
        // Remove all
        await InternshipTypes.destroy({ where: {} });
    });

    it('NoInternshipTypes_200', async () => {
        const RESPONSE = await request(app).delete(`${baseURL}/internshipTypes/10`);
        expect(RESPONSE.status).toBe(200);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).delete(`${baseURL}/internshipTypes/{falseEncoding}`);
        expect(RESPONSE.status).toBe(400);
    });

    it('InternshipTypes_200', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();

        const CREATED = await InternshipTypes.create(VALID_INTERNSHIP_TYPES);

        const RESPONSE = await request(app).delete(`${baseURL}/internshipTypes/${CREATED.id}`);
        expect(RESPONSE.status).toBe(200);
    });
});

describe('GET /internshipTypes/:id/internships', () => {
    beforeEach(async () => {
        // Remove all
        await InternshipTypes.destroy({ where: {} });
    });

    it('NoBusiness_204', async () => {
        const RESPONSE = await request(app).get(`${baseURL}/internshipTypes/10/internships`);
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).get(
            `${baseURL}/internshipTypes/{falseEncoding}/internships`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('InternshipTypes_204_NoLinkedData', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();

        const CREATED = await InternshipTypes.create(VALID_INTERNSHIP_TYPES);
        const RESPONSE = await request(app).get(
            `${baseURL}/internshipTypes/${CREATED.id}/internships`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('InternshipTypes_200_WithLinkedData', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();
        const VALID_INTERNSHIP = defaultInternships();

        VALID_INTERNSHIP_TYPES.internships = [VALID_INTERNSHIP];

        const CREATED_INTERNSHIP_TYPES = await InternshipTypes.create(VALID_INTERNSHIP_TYPES, {
            include: [{ association: InternshipTypes.associations.internships }],
        });

        const RESPONSE = await request(app).get(
            `${baseURL}/internshipTypes/${CREATED_INTERNSHIP_TYPES.id}/internships`,
        );
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toHaveLength(1);
    });
});

describe('POST /internshipTypes/:id/internships/:internship_id/link', () => {
    beforeEach(async () => {
        // Remove all
        await InternshipTypes.destroy({ where: {} });
    });

    it('NoBusiness_204', async () => {
        const RESPONSE = await request(app).post(
            `${baseURL}/internshipTypes/10/internships/20/link`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400_WrongID', async () => {
        const RESPONSE = await request(app).post(
            `${baseURL}/internshipTypes/{falseEncoding}/internships/10/link`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('BadRequest_400_WrongInternshipID', async () => {
        const RESPONSE = await request(app).post(
            `${baseURL}/internshipTypes/10/internships/{falseEncoding}/link`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('InternshipTypes_204_NoInternship', async () => {
        // In this case, we check if link a existing Bussiness and an unexisting internships work
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();

        const CREATED = await InternshipTypes.create(VALID_INTERNSHIP_TYPES);
        const RESPONSE = await request(app).post(
            `${baseURL}/internshipTypes/${CREATED.id}/internships/20/link`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('InternshipTypes_200_WithInternship', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();
        const VALID_INTERNSHIP = defaultInternships();

        let CREATED_INTERNSHIP_TYPES = await InternshipTypes.create(VALID_INTERNSHIP_TYPES);
        const CREATED_INTERNSHIP = await Internships.create(VALID_INTERNSHIP);

        const RESPONSE = await request(app).post(
            `${baseURL}/internshipTypes/${CREATED_INTERNSHIP_TYPES.id}/internships/${CREATED_INTERNSHIP.id}/link`,
        );

        // Should answer 200
        expect(RESPONSE.status).toBe(200);

        // check if business and internship are linked
        CREATED_INTERNSHIP_TYPES = await InternshipTypes.findByPk(CREATED_INTERNSHIP_TYPES.id);

        const internships = await CREATED_INTERNSHIP_TYPES.getInternships();
        expect(internships).toHaveLength(1);
        expect(internships[0].id).toBe(CREATED_INTERNSHIP.id);
    });
});

describe('GET /internshipTypes/:id/campaigns', () => {
    beforeEach(async () => {
        // Remove all
        await InternshipTypes.destroy({ where: {} });
    });

    it('NoBusiness_204', async () => {
        const RESPONSE = await request(app).get(`${baseURL}/internshipTypes/10/campaigns`);
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).get(
            `${baseURL}/internshipTypes/{falseEncoding}/campaigns`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('InternshipTypes_204_NoLinkedData', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();

        const CREATED = await InternshipTypes.create(VALID_INTERNSHIP_TYPES);
        const RESPONSE = await request(app).get(
            `${baseURL}/internshipTypes/${CREATED.id}/campaigns`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('InternshipTypes_200_WithLinkedData', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();
        const VALID_CAMPAIGN = defaultCampaigns();

        let CREATED_INTERNSHIP_TYPES = await InternshipTypes.create(VALID_INTERNSHIP_TYPES);
        const CREATED_CAMPAIGN = await Campaigns.create(VALID_CAMPAIGN);

        await CREATED_INTERNSHIP_TYPES.addCampaign(CREATED_CAMPAIGN.id);
        CREATED_INTERNSHIP_TYPES = await InternshipTypes.findByPk(CREATED_INTERNSHIP_TYPES.id);

        const RESPONSE = await request(app).get(
            `${baseURL}/internshipTypes/${CREATED_INTERNSHIP_TYPES.id}/campaigns`,
        );
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toHaveLength(1);
    });
});

describe('POST /internshipTypes/:id/campaigns/:campaign_id/link', () => {
    beforeEach(async () => {
        // Remove all
        await InternshipTypes.destroy({ where: {} });
    });

    it('NoBusiness_204', async () => {
        const RESPONSE = await request(app).post(`${baseURL}/internshipTypes/10/campaigns/20/link`);
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400_WrongID', async () => {
        const RESPONSE = await request(app).post(
            `${baseURL}/internshipTypes/{falseEncoding}/campaigns/10/link`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('BadRequest_400_WrongCampaignID', async () => {
        const RESPONSE = await request(app).post(
            `${baseURL}/internshipTypes/10/campaigns/{falseEncoding}/link`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('InternshipTypes_204_NoCampaign', async () => {
        // In this case, we check if link a existing Bussiness and an unexisting campaigns work
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();

        const CREATED = await InternshipTypes.create(VALID_INTERNSHIP_TYPES);
        const RESPONSE = await request(app).post(
            `${baseURL}/internshipTypes/${CREATED.id}/campaigns/20/link`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('InternshipTypes_200_WithCampaign', async () => {
        const VALID_INTERNSHIP_TYPES = defaultInternshipTypes();
        const VALID_CAMPAIGN = defaultCampaigns();

        let CREATED_INTERNSHIP_TYPES = await InternshipTypes.create(VALID_INTERNSHIP_TYPES);
        const CREATED_CAMPAIGN = await Campaigns.create(VALID_CAMPAIGN);

        const RESPONSE = await request(app).post(
            `${baseURL}/internshipTypes/${CREATED_INTERNSHIP_TYPES.id}/campaigns/${CREATED_CAMPAIGN.id}/link`,
        );

        // Should answer 200
        expect(RESPONSE.status).toBe(200);

        // check if business and internship are linked
        CREATED_INTERNSHIP_TYPES = await InternshipTypes.findByPk(CREATED_INTERNSHIP_TYPES.id);

        const campaigns = await CREATED_INTERNSHIP_TYPES.getCampaigns();
        expect(campaigns).toHaveLength(1);
        expect(campaigns[0].id).toBe(CREATED_CAMPAIGN.id);
    });
});
