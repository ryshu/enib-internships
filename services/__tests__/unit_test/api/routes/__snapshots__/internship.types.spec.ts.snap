// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /internshipTypes/:id InternshipTypes_200 1`] = `
Object {
  "createdAt": Any<String>,
  "id": Any<Number>,
  "label": "test",
  "updatedAt": Any<String>,
}
`;

exports[`POST /internshipTypes ValidInternshipTypes_200 1`] = `
Object {
  "createdAt": Any<String>,
  "id": Any<Number>,
  "label": "test",
  "updatedAt": Any<String>,
}
`;

exports[`PUT /internshipTypes/:id InternshipTypes_200_NotAnyDataUpdated 1`] = `
Object {
  "createdAt": Any<String>,
  "id": Any<Number>,
  "label": "test",
  "updatedAt": Any<String>,
}
`;

exports[`PUT /internshipTypes/:id InternshipTypes_200_UpdateAllData 1`] = `
Object {
  "createdAt": Any<String>,
  "id": Any<Number>,
  "label": "additional data",
  "updatedAt": Any<String>,
}
`;
